packages/web-frontend/src/sagas.js

```diff
@@ -905,16 +911,56 @@ function* fetchMeasureInfo(mapOverlayId) {
   }
 }
 
-function* fetchMeasureInfoForMeasureChange(action) {
-  yield fetchMeasureInfo(action.mapOverlayId);
+function* fetchMeasureInfoOnlyIfHierarchyIsLoaded(action) {
+  const state = yield select();
+  const { mapOverlayHierarchy } = state.mapOverlayBar;
+  const isMapOverlayHierarchyLoaded = !isMapOverlayHierarchyEmpty(mapOverlayHierarchy);
+  if (isMapOverlayHierarchyLoaded) {
+    if (action.mapOverlayIds.split(',').length > 1) {
+      yield put(onSelectMultipleMapOverlayCheckBox());
+    }
+    yield fetchMeasureInfo();
+  }
 }
 
-function* watchMeasureChange() {
-  yield takeLatest(SET_MAP_OVERLAY, fetchMeasureInfoForMeasureChange);
+function* watchSetMapOverlayChange() {
+  yield takeLatest(SET_MAP_OVERLAY, fetchMeasureInfoOnlyIfHierarchyIsLoaded);
+}
+
+function* callSetMapOverlay(action) {
+  const state = yield select();
+  const currentMapOverlayIds = selectCurrentMapOverlayIds(state);
+  const { multipleMapOverlayCheckbox, mapOverlayHierarchy } = state.mapOverlayBar;
+  const mapOverlayIds = multipleMapOverlayCheckbox
+    ? sortMapOverlayIdsByHierarchyOrder(mapOverlayHierarchy, [
+        ...currentMapOverlayIds,
+        action.mapOverlayId,
+      ])
+    : [action.mapOverlayId];
+  yield put(setMapOverlay(mapOverlayIds.join(',')));
+}
+
+function* watchSelectMapOverlayChange() {
+  yield takeLatest(SELECT_MAP_OVERLAY, callSetMapOverlay);
+}
```

Comment:
That's a good point @kael89 ! `callSetMapOverlay()` will call `SET_MAP_OVERLAY` triggered by `SELECT_MAP_OVERLAY`, I feel our code convention is to keep these similar logics (calling between actions) in `sagas.js` instead of `reducers.js`, which is not the best approach I would say.

However after taking a step back, I felt `selectMapOverlay()` can be a component function within `MapOverlayBar.js` instead of a global action, because it doesn't get called anywhere else. Do you feel that is better approach?

Reply:
Same as `UNSELECT_MAP_OVERLAY `

Reply:
Oh I see what you mean... I like your suggestion for keeping it as a local function, if possible. Actions that call other actions that result in sagas... Things can get really complicated with redux and redux-saga 😄 !